ampd/src/solana/verifier.rs

```diff
@@ -0,0 +1,423 @@
+    sol_tx: &EncodedConfirmedTransactionWithStatusMeta,
+    worker_set: &WorkerSetConfirmation,
+    account_info: &AccountInfo,
+) -> Vote {
+    if !sol_tx
+        .transaction
+        .message
+        .account_keys
+        .contains(source_gateway_address)
+    {
+        info!(
+            tx_id = &worker_set.tx_id,
+            "tx does not contains source_gateway_address"
+        );
+        return Vote::FailedOnChain;
+    }
+
+    if sol_tx.transaction.signatures.is_empty() {
+        info!(tx_id = &worker_set.tx_id, "tx do not contain signatures");
+        return Vote::FailedOnChain;
+    }
+
+    if worker_set.tx_id != sol_tx.transaction.signatures[0] {
+        info!(tx_id = &worker_set.tx_id, "tx_id do not match");
+        return Vote::FailedOnChain;
+    }
+
+    let onchain_operators = match parse_onchain_operators(account_info) {
+        Ok(ops) => ops,
+        Err(err) => {
+            info!(tx_id = &worker_set.tx_id, err = err.to_string());
+            return Vote::FailedOnChain;
+        }
+    };
+
+    match verify_worker_set_operators_data(&worker_set.operators, &onchain_operators) {
+        true => Vote::SucceededOnChain,
+        false => Vote::FailedOnChain,
+    }
+}
+
+fn verify_worker_set_operators_data(
+    ops: &solana_verify_worker_set::Operators,
+    aw_ops: &Operators,
+) -> bool {
+    ops == aw_ops
+}
+
+fn parse_onchain_operators(account_info: &AccountInfo) -> Result<Operators> {
+    if account_info.value.data.len() < 2 {
+        return Err(VerificationError::ParsingError(
+            "Could not find solana account data.".to_string(),
+        ));
+    }
+
+    let account_data = match decode_base64(&account_info.value.data[0]) {
+        Some(data) => data,
+        None => {
+            return Err(VerificationError::ParsingError(
+                "Cannot base64 decode account data.".to_string(),
+            ))
+        }
+    };
+
+    let operators = match borsh::de::from_slice::<Operators>(&account_data) {
+        Ok(ops) => ops,
+        Err(err) => {
+            return Err(VerificationError::ParsingError(format!(
+                "Cannot borsh decode account data: {}",
+                err
+            )))
+        }
+    };
+
+    Ok(operators)
+}
+
+impl PartialEq<auth_weighted::types::operator::Operators> for solana_verify_worker_set::Operators {
+    fn eq(&self, aw_ops: &auth_weighted::types::operator::Operators) -> bool {
+        if self.threshold != *aw_ops.threshold() {
+            return false;
+        }
+
+        // Creating a hashmap for querying the data later. Using a not fixed size key like
+        // Vec<u8> could not be the best. We expect a 33 bytes slice as address. See ['auth_weighted::types::Address::ECDSA_COMPRESSED_PUBKEY_LEN']
+        // So probably we should try to use that after testing the first POC in order to reduce the domain of the key.
+        let addresses_weights_res: Result<HashMap<Vec<u8>, crate::types::U256>> = self
```

Comment:
Should we worry about perf here ?

Reply:
> how many operators we need to compare

Not many I presume, so I think we'll be fine with performance.

Reply:
Is it worth creating a ticket to test this functionality, when we can more easily do that and also mention that we can try to set a fixed key size for the HashMap?